### ISSUE TITLE
`UNION` too sensitive with column ordering

### ISSUE THREAD

**Issue by csubhodeep** [pola-rs/polars#9370]
### Polars version checks

- [X] I have checked that this issue has not already been reported.

- [X] I have confirmed this bug exists on the [latest version](https://pypi.org/project/polars/) of Polars.

### Issue description

The `UNION` clause throws an exception when being applied on two `DataFrame`s that have exactly the same column names (and data types) but different ordering. This relates to #8936 .

### Reproducible example

```python
import polars as pl

df1 = pl.DataFrame({"a": [1], "b": ['x']})
df2 = pl.DataFrame({"b": ['y'], "a": [2]})

sql_ctx = pl.SQLContext(register_globals=True, eager_execution=True)

sql_ctx.execute("""
SELECT * FROM df1
UNION
SELECT * FROM df2
""")

# Output
ShapeError: unable to vstack, column names don't match: "a" and "b"
```

### Expected behavior

```shell
shape: (2, 2)
┌─────┬─────┐
│ a   ┆ b   │
│ --- ┆ --- │
│ i64 ┆ str │
╞═════╪═════╡
│ 1   ┆ x   │
│ 2   ┆ y   │
└─────┴─────┘
```

### Installed versions

<details>

```
--------Version info---------
Polars:      0.18.2
Index type:  UInt32
Platform:    macOS-13.3.1-arm64-arm-64bit
Python:      3.9.16 (main, May 16 2023, 20:00:19) 
[Clang 14.0.3 (clang-1403.0.22.14.1)]

----Optional dependencies----
numpy:       1.24.3
pandas:      2.0.2
pyarrow:     10.0.1
connectorx:  <not installed>
deltalake:   <not installed>
fsspec:      <not installed>
matplotlib:  <not installed>
xlsx2csv:    <not installed>
xlsxwriter:  <not installed>
```

</details>

**Comment by avimallu** [comment]
My understanding of SQL's UNION and UNION ALL operators is that they match by position and not name - am I wrong? A quick check indicated (SQL, not Polars) questions ([1](https://stackoverflow.com/questions/61232740/create-a-union-all-query-when-the-columns-are-in-different-order), [2](https://dba.stackexchange.com/questions/14373/perform-union-all-regardless-of-column-order-in-the-two-joined-tables)) that specifically asked for what you're looking for in the language itself. Some [implementations](https://duckdb.org/docs/sql/query_syntax/setops.html) have a specific `UNION BY NAME` operation to implement this as well.

The strictness is also likely because Polars uses the underlying [concat](https://pola-rs.github.io/polars/py-polars/html/reference/api/polars.concat.html#polars.concat) operation for UNION. I don't think this is a bug - this is likely intended behavior. I would probably raise this as a feature request, since we can use the underlying implementation of concat for `how="diagonal"` for it.

**Comment by csubhodeep** [comment]
@avimallu I agree. Thanks for the clarification. In that case would be great if we could change the label from "bug" to "enhancement" and also add the "sql" label.

**Comment by alexander-beedie** [comment]
> @avimallu I agree. Thanks for the clarification. In that case would be great if we could change the label from "bug" to "enhancement" and also add the "sql" label.

Support for `UNION BY NAME` and `UNION ALL BY NAME` syntax will be available in the upcoming `0.19.8` ✌️
